Support for Activated LoRA (Issue https://github.com/huggingface/peft/issues/2523)

kgreenewald opened the pull request:
Source: https://research.ibm.com/blog/inference-friendly-aloras-lora
Original Github: https://github.com/IBM/activated-lora
Paper: https://arxiv.org/pdf/2504.12397

This PR migrates Activated LoRA (aLoRA) support from a standalone Github (see above) to PEFT itself. 

Note there is also an active PR for vLLM inference support for Activated LoRA: https://github.com/vllm-project/vllm/pull/19710 . There are also collections of aLoRA models on huggingface (in the ibm-granite org), note that these preexisting models run off of the standalone github repo and will be updated to work with this new PEFT feature if merged. 

Description of changes: Activated LoRA is a modification of the LoRA architecture to "activate" the adapter weights only on tokens coming after a specified invocation_string. This fact makes it so that KV values for the string coming before the activation matches KV values for the base model. This allows KV cache for the input to be interchangeable between the base model and adapter model, and allows for major speedups in inference pipelines (e.g. agentic pipelines) that want to use both base models and adapter models. See the paper for detailed exploration of use cases and further elaboration. 

We have tried to make the changes as non-intrusive as possible (but happy to hear any suggestions for how the PR can be improved). The vast majority of changes are in a new folder tuners/alora. Some changes were required to the universally used peft_model.py in order to handle processing of the invocation_string and determination of the activation point. Would be happy to hear any and all suggestions on how to best structure the changes in this file in order to be as nonintrusive as possible.

Other notes: 
* The crux of the changes are really in layer.py. Everything else is simply managing the alora_offsets quantity which defines where the weights start to be activated. This is determined by scanning input strings for the invocation_string defined in the aLoraConfig.
* I believe that aLoRA really only makes sense for CausalLMs, hence I've only implemented this for that model type. 
* Merging doesn't make sense for aLoRA adapters since the weights are not universally applied to all tokens. 
* I used the LoRA code as a starting point, but did not implement various seemingly extra features in that code. If I missed something important definitely let me know!
* As of now, invocation_string should probably start and end with special tokens, to avoid tokenizer issues at the boundary. Open to suggestions on how to make this more general if needed.


![image](https://github.com/user-attachments/assets/59af5b31-1971-46ae-98f9-eb4aaeed7f84)

kgreenewald commented:
> Hey @kgreenewald, thank you so much for your PR!
> 
> I've skimmed the PR and have a few general questions to steer the general implementation before we dig deeper. As far as I understand there are fundamentally two things that we need
> 
>     1. a `forward` method that understands when to activate the adapter and when not to (according to index or binary values, i.e. a mask)
> 
>     2. a way to generate the aforementioned mask from input ids
> 
> 
> for (2) we need to hook into the relevant `PeftModel` instances, there's no way around that but I think that for (1) we don't need a full-fledged PEFT method. I think we can manage with a [LoRA variant](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/variants.py) that just implements the forward call for the `Linear` layer. This way we could reduce the number of changes in this PR by a lot.
> 
> Secondly, I think that it would be best to use the token ids as invocation sequence, not strings. 1) because we would not have to pass the tokenizer around and do tokenization at unexpected places but also because, 2) we will determine the correct token ids at training time once instead of generating them during inference and hiding a possible mixup when using a newer/different tokenizer.
> 
> WDYT?
> 
> Regardless of these comments, I think it would be good to merge rebase to main since #2579 fixes the CI and to implement a few tests. We can start at adding a test case in `tests/test_custom_models.py` which already gives a good test coverage and is generally quite fast to iterate on.

Thanks @githubnemo ! Responses below:

1. These parts 1 & 2 are correct, and the LoraVariant concept is very cool! I actually am worried that this might be _more_ intrusive however -- let me know if you disagree and think this is actually fine! (A) I'd need to change many of the main Lora files in order for the created mask to actually get passed up the chain to the forward method (since (a) the LoraVariant has hardcoded arguments and (b) would need to rewrite the forward hooks). (B) The invocation_tokens need to be stored in the adapter_config.json, I imagine we might not want to create this field for all Lora models? 
---On the other hand, if you believe the masking procedure might be of general interest (even other masking strategies besides the one we currently do, possibly giving users the ability to specify their own masking functions), such that the above downsides are worth dealing with, I'm happy to make the above changes!

2.  "token ids as invocation sequence, not strings": Sounds good, I'll make this change. 

3.  rebase & tests: will do!

githubnemo commented:
@kgreenewald give me a ping when you think you need a second opinion / review!

kgreenewald commented:
> @kgreenewald give me a ping when you think you need a second opinion / review!

@githubnemo I've done the refactor you suggested, Activated LoRA is now a "variant". As suggested, LoraConfig now has fields 
- use_alora
- alora_invocation_string
- alora_invocation_tokens

Everything in this repo is now based off of using alora_invocation_tokens only as you suggested, so no need to pass the tokenizer around. I've retained "alora_invocation_string" in the config for human readability and for external use, e.g. insertion to data as a string if needed.

I also now have some tests in test_custom_models.py , I used the Dora tests as inspiration. 

Let me know what you think, and thanks again!!

githubnemo reviewed (commented):
Thanks a lot for doing the refactor to LoRA variants! This looks quite good already. Good job on adding `kwargs` and shifting the hook-code.

> I've retained "alora_invocation_string" in the config for human 
readability and for external use, e.g. insertion to data as a string if 
needed.

I think we should drop it entirely, alongside `use_alora` so that the only option left is `alora_invocation_tokens`. The tokenizer is only one step away and most likely used in user code anyway, so doing `tokenizer.decode(peft_config.alora_invocation_tokens)` is a lot better compared to storing a 'redundant' string in the config which runs the risk to being outdated due to carelessness IMO.

Regarding testing it is already looking quite good with the custom model tests in place. I'd suggest to remove the `aLoRA does not support XYZ` from `test_custom_models.py` and move it to `testing_common.py` where the actual tests are defined. You'll find that there are already several exceptions defined where you can add aLoRA. It would also make sense to add test-cases to `test_decoder_models.py` and `test_encoder_decoder_models.py` in the same way as`test_custom_models.py`. I think it would also make sense to add an aLoRA-specific test that verifies that the `alora_offsets` are adhered to. You can create a test for this in `tests/test_lora_variants.py`.

kgreenewald commented on src/peft/tuners/lora/config.py:
```diff
@@ -65,8 +65,8 @@ class LoftQConfig:
 
-    loftq_bits: int = field(default=4, metadata={"help": "Quantization bits for LoftQ"})
-    loftq_iter: int = field(default=1, metadata={"help": "Alternating iterations for LoftQ"})
+    loftq_bits: str = field(default=4, metadata={"help": "Quantization bits for LoftQ"})
+    loftq_iter: str = field(default=1, metadata={"help": "Alternating iterations for LoftQ"})
```
yep, will fix!

kgreenewald commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +435,63 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+                    result[:, -offset:, :] = (
+                        result[:, -offset:, :] + lora_B(lora_A(dropout(x[:, -offset:, :]))) * scaling
+                    )
+        else:
```
I'll actually remove the elif: part, now that alora_offsets is computed automatically, this single-entry argument isn't needed I believe

kgreenewald commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +435,63 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+                result[:, :, :] = result[:, :, :] + lora_B(lora_A(dropout(x[:, :, :]))) * scaling
+            else:
+                offset = min(result.shape[1], alora_offsets[0])
+                if offset > 0:
```
I think this also came up in another comment. Right now, -1 means "don't use adapter at all" and "None" means "use as if it were LoRA". Making alora_invocation_tokens != None be the condition for using alora might solve this and remove the need for this distinction, let me figure this out.

kgreenewald replied:
I went with this all None solution.

kgreenewald commented on src/peft/tuners/lora/config.py:
```diff
@@ -300,6 +300,21 @@ class LoraConfig(PeftConfig):
+        alora_invocation_string (`str`):
+            Invocation string for aLoRA (must be present in model inputs). Defaults to None.
+        alora_invocation_tokens (`List[int]`):
+            Tokenized copy of alora_invocation_string for use when tokenizer is not available.
```
Done!

kgreenewald commented on src/peft/peft_model.py:
```diff
@@ -1965,6 +2073,60 @@ def generate(self, *args, **kwargs):
+                        config_ = self.peft_config.get(name)
+                        if config_ and getattr(config_, "use_alora", False):
+                            is_alora_relevant_in_generate = True
+                            break
```
Done!

kgreenewald commented:
@githubnemo Thanks! I went through and made these changes, details follow: 

move alora_offsets logic to variants.py. -> I did this, but it is ever so slightly funky since defining the input to the functions there as PeftModel created circular imports, so I had to have the argument in the function definition listed as "model: nn.module" which is the superclass. If you see a cleaner solution let me know.

"remove the aLoRA does not support XYZ from test_custom_models.py and move it to testing_common.py" -> done

"add test-cases to test_decoder_models.py" -> done

"test_encoder_decoder_models.py" -> I don't think this is applicable, since aLoRA is for CausalLM.

"add an aLoRA-specific test that verifies that the alora_offsets are adhered to" - this mostly happens inside many of the test_decoder_models.py tests. If you'd like, I could maybe make a unit test or two to make sure the alora_offset computation logic is working correctly?

Thanks again for your help, and let me know what you think!

githubnemo reviewed (commented):
Nice, good progress!

> "test_encoder_decoder_models.py" -> I don't think this is applicable, since aLoRA is for CausalLM.

Right, then we should probably add a test specific to aLoRA that an appropriate error is raised when using aLoRA + encoder-decoder model :)

It would be good to have tests for the various warnings that are raised to make sure that they are actually raised. It may be worthwhile to put these specific tests in `test_lora_variants.py`. This could also be the place to have specific tests that make sure that the adapter is actually invoked at the place it is supposed to be invoked. Currently there are only tests that make sure that nothing crashes when offsets are supplied but what really interests us is, was the adapter activated after seeing the invocation tokens and only then. It'd be great if you could add a test for this as well!

Regarding documentation, in the review comments I already mentioned adding a section to the LoRA developer guide. This would also be a good place to describe the benefits of the method and link example code. So, if you have a good example in mind how to show-case aLoRA, you can add it to the `examples` directory and link it in the documentation. Having a ready-made example showcasing the technique could be quite helpful for users.

A minor thing but spread throughout: There are several `nn.module` typos, it is `nn.Module`.

Keep up the good work :)

kgreenewald commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +438,203 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+
+def calculate_alora_offsets(
+    peft_config: PeftConfig, active_adapter: str, input_ids: torch.Tensor, adapter_names: Optional[list[str]] = None
+) -> list[int]:
```
Docstrings added for all of these

kgreenewald commented on src/peft/peft_model.py:
```diff
@@ -1825,7 +1826,10 @@ def forward(
         peft_config = self.active_peft_config
+
         if not peft_config.is_prompt_learning:
+            # For aLoRA
```
done!

kgreenewald commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +438,203 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+
+def get_alora_offsets_for_forward(model: nn.module, input_ids: torch.Tensor, inputs_embeds: torch.Tensor, **kwargs):
+    adapter_names_for_offset_calc = kwargs.get("adapter_names", None)
+    if is_alora_relevant_in_batch(model, adapter_names_for_offset_calc):
```
Done!

kgreenewald commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +438,203 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+                ):  # Should have been caught by current_input_ids
+                    bs = kwargs["input_ids"].shape[0]
+
+                kwargs["alora_offsets"] = [None] * bs
```
Correct, fixed

kgreenewald commented on tests/testing_common.py:
```diff
@@ -1019,11 +1036,11 @@ def _test_mixed_adapter_batches(self, model_id, config_cls, config_kwargs):
             logits_mixed = model.generate(**dummy_input, return_dict_in_generate=True, output_scores=True).scores[0]
-
+        # print(output_adapter0[1::3])
+        # print(output_mixed[1::3])
```
done

kgreenewald commented:
@githubnemo Comments incorporated, as well as new docs, new alora_offsets related tests, and a new example. 

FYI - seems like the "_mixed_batch_forward" wasn't actually implemented for lora variants (no warnings or comments). I went ahead and implemented it, seemed pretty straightforward but happy to revert if there's an underlying reason.

kgreenewald commented on docs/source/developer_guides/lora.md:
````diff
@@ -173,6 +173,51 @@ from peft import LoraConfig
+invocation_string = "placeholder"
+alora_invocation_tokens = tokenizer.encode(invocation_string, add_special_tokens=False).
+```
+where the tokenizer is the tokenizer for the base model.
````
Sounds good, done!

kgreenewald commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +438,232 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+            elif (
+                "input_ids" in kwargs and kwargs["input_ids"] is not None
+            ):  # Should have been caught by current_input_ids
+                bs = kwargs["input_ids"].shape[0]
```
Sorry, fixed!

githubnemo commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +438,232 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+            if best_match_start_idx != -1:
+                offset_val = seq_len - best_match_start_idx + 1
+                alora_offsets[i] = offset_val if offset_val > 0 else None
+            else:  # Invocation sequence not found in input
+                warnings.warn(
+                    f"Could not find alora_invocation_tokens for specified aLoRA adapter in the "
+                    f"following instance"
+                    f"{sequence}"
+                    f"Invocation tokens: {current_invocation_ids_tensor} \n"
+                    f"Defaulting to base model. "
+                )
+                alora_offsets[i] = None
```
I see, let's remove it then. Thanks for explaining!

githubnemo commented on docs/source/developer_guides/lora.md:
````diff
@@ -173,6 +173,51 @@ from peft import LoraConfig
+alora_invocation_tokens = tokenizer.encode(invocation_string, add_special_tokens=False).
+```
+
+**Note** If using custom strings for the invocation string, make sure that the start and end of the string are special tokens to avoid issues with tokenization at the boundaries.
````
Alright, makes sense. I think it's worthwhile to be more elaborate about this in the docs, similar to the explanation you gave me here.

kgreenewald replied:
added to the docs!

githubnemo commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +438,232 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+        if x.dim() == 2:
+            # If x is 2-dimensional (unusual but comes up in certain tests), this means that for all inputs,
+            # there is only 1 token position being processed and we should adapt its weights.
+            result = result + lora_B(lora_A(dropout(x))) * scaling
```
Can you tell me which tests are relevant to this?

kgreenewald replied:
Every alora test in test_custom_modules.py, but no other files I believe

kgreenewald commented on tests/test_lora_variants.py:
```diff
@@ -124,3 +156,50 @@ def test_dora_params_have_gradients(self):
+        lora_out = lora_model(X=input_ids, **kwargs)
+    start = input_ids.shape[1] - kwargs["alora_offsets"][0]
+    assert torch.allclose(lora_out[:, :start], base_out[:, :start])
+    assert not torch.allclose(lora_out[:, start:], base_out[:, start:])
```
done

kgreenewald commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +438,232 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+            )
+            for i in range(len(calculated_offsets)):
+                if calculated_offsets[i] is not None:
+                    calculated_offsets[i] -= 1
```
Fixed, so this isn't here anymore.

kgreenewald commented on tests/test_lora_variants.py:
```diff
@@ -124,3 +156,50 @@ def test_dora_params_have_gradients(self):
+
+    offsets = calculate_alora_offsets(peft_config, "a1", input_ids, adapter_names=["a1", "a2"])
+
+    assert offsets == [input_ids.shape[1] - 2 + 1, input_ids.shape[1] - 2 + 1]
```
done

kgreenewald commented on src/peft/tuners/lora/bnb.py:
```diff
@@ -178,13 +185,14 @@ def _mixed_batch_forward(
             for adapter in unique_adapters:
                 sub_batch_indices_list.append([index for index, item in enumerate(adapter_names) if item == adapter])
-
+            alora_offsets = variant_kwargs.get("alora_offsets", None)
```
done

kgreenewald commented:
> Cool, I really liked the docs, they read nicely.
> 
> Some smaller comments left.
> 
> It might be worthwhile to also add aLoRA to the `VARIANT_MAP` / `TEST_CASES` matrix in `test_lora_variants.py` for added coverage regarding variant application.
> 
> > FYI - seems like the "_mixed_batch_forward" wasn't actually implemented for lora variants (no warnings or comments). I went ahead and implemented it, seemed pretty straightforward but happy to revert if there's an underlying reason.
> 
> Thanks! Looks good to me :)

Thanks! I've done these changes, modulo the remaining comment about 2D vs. 3D inputs.

githubnemo reviewed (commented):
I think except for a few missing / incomplete tests we're almost done.

Once these tests are done I'd let the CI run and if everything's green we can merge this :)

Make sure to keep up with main and run `make style`.

githubnemo commented:
> > I think except for a few missing / incomplete tests we're almost done.
> > Since this PR also provides a bitsandbytes implementation, make sure to add a test in `test_gpu_examples.py` (similar to the Eva test, roughly testing inference with bitsandbytes on one of the smaller models there, like OPT). Once these tests are done I'd let the CI run and if everything's green we can merge this :)
> > Make sure to keep up with main and run `make style`.
> 
> @githubnemo Great!! I went through and did these changes, including going ahead and doing the "mask" refactor for the alora forward in variants.py.
> 
> EDIT - how should I get these CI checks working right on github? Seems there's some token etc I need to provide? Sorry if I'm missing something obvious :)

No worries, the CI is configured that only after maintainer review the tests are run (security reasons). You can run these tests locally as well:

- `make style` from the root directory for linting / formatting
- `pytest tests/` to run all tests
- `pytest -k alora tests` to run tests with 'alora' in their name (might not be exhaustive)

I'll trigger a CI run now. If runs through I'll do a review, otherwise I'll give you a ping :)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2609). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

githubnemo reviewed (commented):
Good work! Three little comments, everything else looks fine now.

One thing I'm still unsure about is caching. If I understand correctly - correct me if I'm wrong - we're not supposed to cache the KV values that occur when the adapter is active. This would only matter caches that are shared between `generate` invocations since we don't return to the base model after adapter invocation.

I think we would still need to handle the case in `PeftModel.generate` of receiving a [`Cache`](https://huggingface.co/docs/transformers/v4.56.0/en/internal/generation_utils#transformers.Cache) class in `kwargs['past_key_values']`. I think we'd need to truncate the KV cache after the call to `base_model.generate` to the point before the adapter activation. But I'm not  yet sure how to do this properly. If you've got any ideas, please let me know. I'd be fine to raise an error instead for the time being that external caching + aLoRA is not supported right now.

kgreenewald commented:
> Good work! Three little comments, everything else looks fine now.
> 
> One thing I'm still unsure about is caching. If I understand correctly - correct me if I'm wrong - we're not supposed to cache the KV values that occur when the adapter is active. This would only matter caches that are shared between `generate` invocations since we don't return to the base model after adapter invocation.
> 
> I think we would still need to handle the case in `PeftModel.generate` of receiving a [`Cache`](https://huggingface.co/docs/transformers/v4.56.0/en/internal/generation_utils#transformers.Cache) class in `kwargs['past_key_values']`. I think we'd need to truncate the KV cache after the call to `base_model.generate` to the point before the adapter activation. But I'm not yet sure how to do this properly. If you've got any ideas, please let me know. I'd be fine to raise an error instead for the time being that external caching + aLoRA is not supported right now.

@githubnemo This is an interesting and subtle question - it is definitely possible/desired to use the cache for generation, but as you point out there are a lot of ways to do things incorrectly. Here's a rough example of how you can do things in a multi-step pipeline where first a "safety" adapter is called on the input, second the base model generates from the input, then two adapters check that same input. Note that each step re-uses cache. For the last checks, feeding in the cache from the base model generation just works and is straightforward. But for the first "safety" turn, the base model hasn't done anything yet, but we want to create cache that the base model call can use. So I first do a prefill operation on the input using the base model to create a input-cache, which is then used both by "safety" and "base".

NOTE - this code snippet is from a prior implementation, you might notice some slight differences like passing in alora_offsets directly. The cache stuff should still apply though.

```python
prompt_cache = DynamicCache()
input_text = tokenizer.apply_chat_template(question_chat,tokenize=False,add_generation_prompt=False) #+ safety_prompt
inputs = tokenizer(input_text, return_tensors="pt")
with model_alora.disable_adapter():
    with torch.no_grad():
        prompt_cache = model_alora(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), past_key_values=prompt_cache).past_key_values 

# Generate safety exception
input_safety, alora_offsets = tokenize_alora(tokenizer,input_text, safety_prompt)

past_key_values = copy.deepcopy(prompt_cache)
output = model_alora.generate(input_safety["input_ids"].to(device), attention_mask=input_safety["attention_mask"].to(device), use_cache=True, max_new_tokens=10, return_dict_in_generate=True, past_key_values = past_key_values, alora_offsets = alora_offsets)

output_text = tokenizer.decode(output.sequences[0])

answer = output_text.split(safety_prompt)[-1]
print("Safety: " + answer)

question_chat = [
    {
        "role": "system",
        "content": system_prompt
    },
    {
        "role": "user",
        "content": question
    },
]

# Generate answer with base
input_text = tokenizer.apply_chat_template(question_chat,tokenize=False,add_generation_prompt=True)
inputs = tokenizer(input_text, return_tensors="pt")
with model_alora.disable_adapter():
    output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=160, past_key_values = prompt_cache, return_dict_in_generate=True)
output_text = tokenizer.decode(output.sequences[0])
prompt_cache = output.past_key_values
answer = output_text.split("assistant<|end_of_role|>")[-1]
print("Answer: " + answer)

# Generate certainty score
uq_generation_prompt = "<|start_of_role|>certainty<|end_of_role|>"
uq_chat = [
    {
        "role": "system",
        "content": system_prompt
    },
    {
        "role": "user",
        "content": question
    },
    {
        "role": "assistant",
        "content": answer
    },
]

uq_text = tokenizer.apply_chat_template(uq_chat,tokenize=False) #+ uq_generation_prompt
inputs, alora_offsets = tokenize_alora(tokenizer,uq_text, uq_generation_prompt)

model_alora.set_adapter("certainty")
answer_KV = copy.deepcopy(prompt_cache)
output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=6, past_key_values=answer_KV,return_dict_in_generate=True, alora_offsets = alora_offsets)
output_text = tokenizer.decode(output.sequences[0])
print("Certainty: " + output_text.split("certainty<|end_of_role|>")[-1])

#Hallucination
model_alora.set_adapter("hallucination")
hall_prompt = "<|start_of_role|>hallucination<|end_of_role|>"
uq_text = tokenizer.apply_chat_template(uq_chat,tokenize=False) 
#inputs = tokenizer(uq_text, return_tensors="pt")
inputs, alora_offsets = tokenize_alora(tokenizer,uq_text, hall_prompt)
output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=5,past_key_values=copy.deepcopy(prompt_cache),return_dict_in_generate=True, alora_offsets = alora_offsets)
output_text = tokenizer.decode(output.sequences[0])
print("Hallucination: " + output_text.split("hallucination<|end_of_role|>")[-1])
```


Regarding truncating the created cache in .generate, I agree that this is probably the right thing to do in most cases, but from past attempts I recall that doing the truncation is not terribly well supported? I've done it, but it involved converting to legacy cache and doing operations on the data itself. This hack probably wouldn't be general enough, e.g. with Mamba caches and whatnot. 

For how to handle this - I think it's pretty important to still allow for cache to be provided as an argument. Maybe some possible solutions (?): 1. leave as-is but document this point, 2. deepcopy the input cache and just return that (and document this), 3. give warnings for either of the above cases whenever cache is passed in, stating that the resulting cache will be hybrid (possibly becoming spammy). 4 (hardest, likely messy) try to figure out how to truncate the cache.

kgreenewald commented on docs/source/developer_guides/lora.md:
```diff
@@ -173,6 +173,53 @@ from peft import LoraConfig
+
+Activated LoRA (aLoRA) is a low rank adapter architecture for Causal LMs that allows for reusing existing base model KV cache for more efficient inference. This approach is best suited for inference pipelines which rely on the base model for most tasks/generations, but use aLoRA adapter(s) to perform specialized task(s) within the chain. For example, checking or correcting generated outputs of th ...
+
+This technique scans for the last occurence of an invocation sequence (`alora_invocation_tokens`) in each input (this can be as short as 1 token), and activates the adapter weights on tokens starting 1 token after the beginning of the invocation sequence. Weights on prior tokens are left un-adapted -- making the cache for those tokens interchangeable with base model cache due to the causal attent ...
```
Added this!

githubnemo commented:
> @githubnemo This is an interesting and subtle question - it is definitely possible/desired to use the cache for generation, but as you point out there are a lot of ways to do things incorrectly. Here's a rough example of how you can do things in a multi-step pipeline where first a "safety" adapter is called on the input, second the base model generates from the input, then two adapters check that same input. Note that each step re-uses cache. For the last checks, feeding in the cache from the base model generation just works and is straightforward. But for the first "safety" turn, the base model hasn't done anything yet, but we want to create cache that the base model call can use. So I first do a prefill operation on the input using the base model to create a input-cache, which is then used both by "safety" and "base".
> 
> NOTE - this code snippet is from a prior implementation, you might notice some slight differences like passing in alora_offsets directly. The cache stuff should still apply though.
> 
> ```python
> prompt_cache = DynamicCache()
> input_text = tokenizer.apply_chat_template(question_chat,tokenize=False,add_generation_prompt=False) #+ safety_prompt
> inputs = tokenizer(input_text, return_tensors="pt")
> with model_alora.disable_adapter():
>     with torch.no_grad():
>         prompt_cache = model_alora(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), past_key_values=prompt_cache).past_key_values 
> 
> # Generate safety exception
> input_safety, alora_offsets = tokenize_alora(tokenizer,input_text, safety_prompt)
> 
> past_key_values = copy.deepcopy(prompt_cache)
> output = model_alora.generate(input_safety["input_ids"].to(device), attention_mask=input_safety["attention_mask"].to(device), use_cache=True, max_new_tokens=10, return_dict_in_generate=True, past_key_values = past_key_values, alora_offsets = alora_offsets)
> 
> output_text = tokenizer.decode(output.sequences[0])
> 
> answer = output_text.split(safety_prompt)[-1]
> print("Safety: " + answer)
> 
> question_chat = [
>     {
>         "role": "system",
>         "content": system_prompt
>     },
>     {
>         "role": "user",
>         "content": question
>     },
> ]
> 
> # Generate answer with base
> input_text = tokenizer.apply_chat_template(question_chat,tokenize=False,add_generation_prompt=True)
> inputs = tokenizer(input_text, return_tensors="pt")
> with model_alora.disable_adapter():
>     output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=160, past_key_values = prompt_cache, return_dict_in_generate=True)
> output_text = tokenizer.decode(output.sequences[0])
> prompt_cache = output.past_key_values
> answer = output_text.split("assistant<|end_of_role|>")[-1]
> print("Answer: " + answer)
> 
> # Generate certainty score
> uq_generation_prompt = "<|start_of_role|>certainty<|end_of_role|>"
> uq_chat = [
>     {
>         "role": "system",
>         "content": system_prompt
>     },
>     {
>         "role": "user",
>         "content": question
>     },
>     {
>         "role": "assistant",
>         "content": answer
>     },
> ]
> 
> uq_text = tokenizer.apply_chat_template(uq_chat,tokenize=False) #+ uq_generation_prompt
> inputs, alora_offsets = tokenize_alora(tokenizer,uq_text, uq_generation_prompt)
> 
> model_alora.set_adapter("certainty")
> answer_KV = copy.deepcopy(prompt_cache)
> output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=6, past_key_values=answer_KV,return_dict_in_generate=True, alora_offsets = alora_offsets)
> output_text = tokenizer.decode(output.sequences[0])
> print("Certainty: " + output_text.split("certainty<|end_of_role|>")[-1])
> 
> #Hallucination
> model_alora.set_adapter("hallucination")
> hall_prompt = "<|start_of_role|>hallucination<|end_of_role|>"
> uq_text = tokenizer.apply_chat_template(uq_chat,tokenize=False) 
> #inputs = tokenizer(uq_text, return_tensors="pt")
> inputs, alora_offsets = tokenize_alora(tokenizer,uq_text, hall_prompt)
> output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=5,past_key_values=copy.deepcopy(prompt_cache),return_dict_in_generate=True, alora_offsets = alora_offsets)
> output_text = tokenizer.decode(output.sequences[0])
> print("Hallucination: " + output_text.split("hallucination<|end_of_role|>")[-1])
> ```
> 
> Regarding truncating the created cache in .generate, I agree that this is probably the right thing to do in most cases, but from past attempts I recall that doing the truncation is not terribly well supported? I've done it, but it involved converting to legacy cache and doing operations on the data itself. This hack probably wouldn't be general enough, e.g. with Mamba caches and whatnot.
> 
> For how to handle this - I think it's pretty important to still allow for cache to be provided as an argument. Maybe some possible solutions (?): 1. leave as-is but document this point, 2. deepcopy the input cache and just return that (and document this), 3. give warnings for either of the above cases whenever cache is passed in, stating that the resulting cache will be hybrid (possibly becoming spammy). 4 (hardest, likely messy) try to figure out how to truncate the cache.

Thanks for the detailed answer.

I agree that (4) is definitely the hardest, I spent quite some time to find a good, common way and wasn't able to. While I like (2) I'm a bit wary of possible side-effects, so (1) - documenting, possibly with a slightly cleaned/compacted version of the code snippet you provided, seems to me the best option at the moment.

kgreenewald commented:
> > @githubnemo This is an interesting and subtle question - it is definitely possible/desired to use the cache for generation, but as you point out there are a lot of ways to do things incorrectly. Here's a rough example of how you can do things in a multi-step pipeline where first a "safety" adapter is called on the input, second the base model generates from the input, then two adapters check that same input. Note that each step re-uses cache. For the last checks, feeding in the cache from the base model generation just works and is straightforward. But for the first "safety" turn, the base model hasn't done anything yet, but we want to create cache that the base model call can use. So I first do a prefill operation on the input using the base model to create a input-cache, which is then used both by "safety" and "base".
> > NOTE - this code snippet is from a prior implementation, you might notice some slight differences like passing in alora_offsets directly. The cache stuff should still apply though.
> > ```python
> > prompt_cache = DynamicCache()
> > input_text = tokenizer.apply_chat_template(question_chat,tokenize=False,add_generation_prompt=False) #+ safety_prompt
> > inputs = tokenizer(input_text, return_tensors="pt")
> > with model_alora.disable_adapter():
> >     with torch.no_grad():
> >         prompt_cache = model_alora(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), past_key_values=prompt_cache).past_key_values 
> > 
> > # Generate safety exception
> > input_safety, alora_offsets = tokenize_alora(tokenizer,input_text, safety_prompt)
> > 
> > past_key_values = copy.deepcopy(prompt_cache)
> > output = model_alora.generate(input_safety["input_ids"].to(device), attention_mask=input_safety["attention_mask"].to(device), use_cache=True, max_new_tokens=10, return_dict_in_generate=True, past_key_values = past_key_values, alora_offsets = alora_offsets)
> > 
> > output_text = tokenizer.decode(output.sequences[0])
> > 
> > answer = output_text.split(safety_prompt)[-1]
> > print("Safety: " + answer)
> > 
> > question_chat = [
> >     {
> >         "role": "system",
> >         "content": system_prompt
> >     },
> >     {
> >         "role": "user",
> >         "content": question
> >     },
> > ]
> > 
> > # Generate answer with base
> > input_text = tokenizer.apply_chat_template(question_chat,tokenize=False,add_generation_prompt=True)
> > inputs = tokenizer(input_text, return_tensors="pt")
> > with model_alora.disable_adapter():
> >     output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=160, past_key_values = prompt_cache, return_dict_in_generate=True)
> > output_text = tokenizer.decode(output.sequences[0])
> > prompt_cache = output.past_key_values
> > answer = output_text.split("assistant<|end_of_role|>")[-1]
> > print("Answer: " + answer)
> > 
> > # Generate certainty score
> > uq_generation_prompt = "<|start_of_role|>certainty<|end_of_role|>"
> > uq_chat = [
> >     {
> >         "role": "system",
> >         "content": system_prompt
> >     },
> >     {
> >         "role": "user",
> >         "content": question
> >     },
> >     {
> >         "role": "assistant",
> >         "content": answer
> >     },
> > ]
> > 
> > uq_text = tokenizer.apply_chat_template(uq_chat,tokenize=False) #+ uq_generation_prompt
> > inputs, alora_offsets = tokenize_alora(tokenizer,uq_text, uq_generation_prompt)
> > 
> > model_alora.set_adapter("certainty")
> > answer_KV = copy.deepcopy(prompt_cache)
> > output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=6, past_key_values=answer_KV,return_dict_in_generate=True, alora_offsets = alora_offsets)
> > output_text = tokenizer.decode(output.sequences[0])
> > print("Certainty: " + output_text.split("certainty<|end_of_role|>")[-1])
> > 
> > #Hallucination
> > model_alora.set_adapter("hallucination")
> > hall_prompt = "<|start_of_role|>hallucination<|end_of_role|>"
> > uq_text = tokenizer.apply_chat_template(uq_chat,tokenize=False) 
> > #inputs = tokenizer(uq_text, return_tensors="pt")
> > inputs, alora_offsets = tokenize_alora(tokenizer,uq_text, hall_prompt)
> > output = model_alora.generate(inputs["input_ids"].to(device), attention_mask=inputs["attention_mask"].to(device), max_new_tokens=5,past_key_values=copy.deepcopy(prompt_cache),return_dict_in_generate=True, alora_offsets = alora_offsets)
> > output_text = tokenizer.decode(output.sequences[0])
> > print("Hallucination: " + output_text.split("hallucination<|end_of_role|>")[-1])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Regarding truncating the created cache in .generate, I agree that this is probably the right thing to do in most cases, but from past attempts I recall that doing the truncation is not terribly well supported? I've done it, but it involved converting to legacy cache and doing operations on the data itself. This hack probably wouldn't be general enough, e.g. with Mamba caches and whatnot.
> > For how to handle this - I think it's pretty important to still allow for cache to be provided as an argument. Maybe some possible solutions (?): 1. leave as-is but document this point, 2. deepcopy the input cache and just return that (and document this), 3. give warnings for either of the above cases whenever cache is passed in, stating that the resulting cache will be hybrid (possibly becoming spammy). 4 (hardest, likely messy) try to figure out how to truncate the cache.
> 
> Thanks for the detailed answer.
> 
> I agree that (4) is definitely the hardest, I spent quite some time to find a good, common way and wasn't able to. While I like (2) I'm a bit wary of possible side-effects, so (1) - documenting, possibly with a slightly cleaned/compacted version of the code snippet you provided, seems to me the best option at the moment.

@githubnemo Ok, done! Put a discussion in lora.md in the developer guides. Tried to make the code snippets short, but definitely let me know if the style of these should be improved. I also have a sentence here mentioning that there are PRs to vLLM and llama.cpp that avoid having to do all this manual cache handling, hope that's ok!

githubnemo reviewed (commented):
Cool! I left a few suggestions regarding the cache example. Otherwise the PR LGTM.

Once that section is done we can merge :)

githubnemo commented:
This looks good to go to me. Once the CI is green, I'll merge :)

Thank you for seeing this implementation through!
